scripts/base/files/x509/main.zeek

```diff
@@ -50,13 +110,47 @@ event zeek_init() &priority=5
 	Files::register_for_mime_type(Files::ANALYZER_SHA1, "application/x-x509-user-cert");
 	Files::register_for_mime_type(Files::ANALYZER_SHA1, "application/x-x509-ca-cert");
 	Files::register_for_mime_type(Files::ANALYZER_SHA1, "application/pkix-cert");
+
+	# SHA256 is used by us to determine which certificates to cache.
+	Files::register_for_mime_type(Files::ANALYZER_SHA256, "application/x-x509-user-cert");
+	Files::register_for_mime_type(Files::ANALYZER_SHA256, "application/x-x509-ca-cert");
+	Files::register_for_mime_type(Files::ANALYZER_SHA256, "application/pkix-cert");
+
+	x509_set_certificate_cache(certificate_cache);
+	x509_set_certificate_cache_hit_callback(x509_certificate_cache_replay);
 	}
 
-redef record Files::Info += {
-	## Information about X509 certificates. This is used to keep
-	## certificate information until all events have been received.
-	x509: X509::Info &optional;
-};
+hook x509_certificate_cache_replay(f: fa_file, sha256: string)
+	{
+	# we encountered a cached cert. The X509 analyzer will skip it. Let's raise all the events that it typically
+	# raises by ourselfes.
+
+	# first - let's checked if it already has an x509 record. That would mean that someone raised the file_hash event
+	# several times for the certificate - in which case we bail out.
+	if ( f$info?$x509 )
+		return;
+
+	local e = certificate_cache[sha256];
```

Comment:
Also just thought: this lookup already happened internally, so could potentially pass the result through via `hook` argument and avoid the 2nd lookup.  May use type `any` and cast if you want to keep API generic.

Reply:
>     * Any idea of what the performance benefit vs. memory tradeoff is in either some real or benchmark measurement to confirm these changes have intended results?

Honestly, only a bit. The expected simultaneous number of certificates on larger links is a few 1,000. If we go with 5,000 and an average cert size of 50kb, this uses ~250mb of ram per worker process. Which is a not insignificant value.

For this you reduce the OpenSSL operations to basically 0 - and the number of allocations for certificates also goes way down. That being said - only testing this in a real-world environment will show if it is really worth it - I, however, expect that it is :)

This also opens the door to more drastic optimizations for sites that have really high TLS/certificate traffic - which can now just suppress events for duplicate certs.

Thanks for all the other suggestions - I will go and clean this up.